File: android/app/src/main/java/com/airbridge/app/domain/NotificationPayload.kt
package com.airbridge.app.domain

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class NotificationEvent {
    @SerialName("posted")
    POSTED,

    @SerialName("updated")
    UPDATED,

    @SerialName("removed")
    REMOVED,
}

@Serializable
data class NotificationImagePayload(
    @SerialName("mime_type")
    val mimeType: String,
    @SerialName("data_base64")
    val dataBase64: String,
    val width: Int,
    val height: Int,
)

@Serializable
data class NotificationAssetPayload(
    @SerialName("app_icon")
    val appIcon: NotificationImagePayload? = null,
    @SerialName("large_icon")
    val largeIcon: NotificationImagePayload? = null,
    @SerialName("hero_image")
    val heroImage: NotificationImagePayload? = null,
)

@Serializable
data class NotificationPayload(
    @SerialName("schema_version")
    val schemaVersion: Int = 2,
    val event: NotificationEvent,
    @SerialName("notification_id")
    val notificationId: String,
    @SerialName("package_name")
    val packageName: String,
    @SerialName("app_name")
    val appName: String,
    val title: String? = null,
    val subtitle: String? = null,
    val body: String? = null,
    @SerialName("posted_at")
    val postedAt: String,
    @SerialName("observed_at")
    val observedAt: String,
    @SerialName("is_ongoing")
    val isOngoing: Boolean,
    val category: String? = null,
    @SerialName("channel_id")
    val channelId: String? = null,
    @SerialName("channel_name")
    val channelName: String? = null,
    val assets: NotificationAssetPayload? = null,
)

File: mac/AirBridgeMac/Domain/NotificationPayload.swift
import Foundation

enum NotificationEventType: String, Codable {
    case posted
    case updated
    case removed
}

struct NotificationPayload: Codable, Equatable {
    let schemaVersion: Int
    let event: NotificationEventType
    let notificationID: String
    let packageName: String
    let appName: String
    let title: String?
    let subtitle: String?
    let body: String?
    let postedAt: Date
    let observedAt: Date
    let isOngoing: Bool
    let category: String?
    let channelID: String?
    let channelName: String?
    let assets: NotificationAssetPayload?

    var remoteIdentifier: String {
        notificationID
    }

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case event
        case notificationID = "notification_id"
        case packageName = "package_name"
        case appName = "app_name"
        case title
        case subtitle
        case body
        case postedAt = "posted_at"
        case observedAt = "observed_at"
        case isOngoing = "is_ongoing"
        case category
        case channelID = "channel_id"
        case channelName = "channel_name"
        case assets
    }
}

struct NotificationAssetPayload: Codable, Equatable {
    let appIcon: NotificationImagePayload?
    let largeIcon: NotificationImagePayload?
    let heroImage: NotificationImagePayload?

    enum CodingKeys: String, CodingKey {
        case appIcon = "app_icon"
        case largeIcon = "large_icon"
        case heroImage = "hero_image"
    }
}

struct NotificationImagePayload: Codable, Equatable {
    let mimeType: String
    let dataBase64: String
    let width: Int
    let height: Int

    enum CodingKeys: String, CodingKey {
        case mimeType = "mime_type"
        case dataBase64 = "data_base64"
        case width
        case height
    }
}
